snuba/optimize.py

```diff
@@ -126,15 +128,68 @@ def get_partitions_to_optimize(
     return parts
 
 
-def optimize_partitions(
+def _subdivide_parts(
+    parts: Sequence[util.Part], number_of_subdivisions: int
+) -> Sequence[Sequence[util.Part]]:
+    """
+    Subdivide a list of parts into number_of_subdivisions lists of parts so that optimize can be executed
+    on different parts list by different threads.
+    """
+
+    sorted_parts = sorted(parts, key=lambda part: part.date, reverse=True)
+    output: MutableSequence[Sequence[util.Part]] = []
+
+    for i in range(number_of_subdivisions):
+        output.append(sorted_parts[i::number_of_subdivisions])
+
+    return output
+
+
+def optimize_partition_runner(
     clickhouse: ClickhousePool,
     database: str,
     table: str,
     parts: Sequence[util.Part],
     ignore_cutoff: bool,
+    parallel: int,
     clickhouse_host: Optional[str] = None,
 ) -> None:
+    if parallel <= 1:
+        optimize_partitions(
+            clickhouse, database, table, parts, ignore_cutoff, clickhouse_host
+        )
+    else:
+        divided_parts = _subdivide_parts(parts, parallel)
+        threads: MutableSequence[threading.Thread] = []
+        for i in range(0, parallel):
+            threads.append(
+                threading.Thread(
+                    target=optimize_partitions,
+                    args=(
+                        clickhouse,
+                        database,
+                        table,
+                        divided_parts[i],
+                        ignore_cutoff,
+                        clickhouse_host,
+                    ),
+                )
+            )
 
+            threads[i].start()
+
+        for i in range(0, parallel):
+            threads[i].join()
```

Comment:
I don't think we can do this. 
We cannot run parallel optimize for the entire duration of the job. As the process will end too close to peak load time.
We need to set a time boundary where we start running in parallel and where we stop running parallel and go back to serial. 

Which means this code needs to be arranged differently, having a number of threads,  then scheduling the next part on a thread as soon as we can and change the number of available threads depending what time it is.